### PR TITLE
[@task] allow initial value

### PR DESCRIPTION
**Pull request by taylor-vann**
Currently the `initial value` of `Task` is assumed to be `undefined`.
This is the default behavior. When a Task is "re-run", `value` is set as `undefined`.

This PR adds an `initialValue` to the `TaskConfig` interface.
When no `intialValue` is set, the behavior described above is preserved.

When `initialValue` is set, the value is only updated when the task completes. As in, the initial value will not be set in future `PENDING` states.

### CODE REVIEW

**Comment by taylor-vann** [comment]
related to: #2367

File: packages/labs/task/src/task.ts [hunk 1]
```diff
@@ -105,6 +106,7 @@ export class Task<T extends [...unknown[]] = any, R = any> {
   private _getArgs?: ArgsFunction<T>;
   private _callId = 0;
   private _host: ReactiveControllerHost;
+  private _intialValue?: R;
```
Comment (mohe2015) [inline packages/labs/task/src/task.ts:109]: spelling mistake

File: packages/labs/task/src/task.ts [hunk 2]
```diff
@@ -34,16 +34,17 @@ export const initialState = Symbol();
 export type TaskStatus = typeof TaskStatus[keyof typeof TaskStatus];
 
 export type StatusRenderer<R> = {
-  initial?: () => unknown;
-  pending?: () => unknown;
-  complete?: (value: R) => unknown;
+  initial?: (value?: R) => unknown;
+  pending?: (value?: R) => unknown;
+  complete?: (value?: R) => unknown;
```
Comment (mohe2015) [inline packages/labs/task/src/task.ts:39]: I don't really like that the `complete` value is now optional.

**Comment by mohe2015** [comment]
What is a concrete use case for this? I don't fully understand that yet.

File: packages/labs/task/src/task.ts [hunk 3]
```diff
@@ -193,8 +197,8 @@ export class Task<T extends [...unknown[]] = any, R = any> {
       });
     }
     this.status = TaskStatus.PENDING;
+    this._value = this._intialValue;
```
Comment (sorvell) [inline packages/labs/task/src/task.ts:200]: I thought we decided to the value stick around and not reset it at all?

File: packages/labs/task/src/task.ts (hunk 2 above)
Comment (sorvell) [reply packages/labs/task/src/task.ts:39]: Is making `value` in `complete` optional related to allowing an initial value? If not, let's not change that.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/labs/task/src/task.ts (hunk 1 above)
Comment (taylor-vann) [reply packages/labs/task/src/task.ts:109]: thank you! <3

File: packages/labs/task/src/task.ts (hunk 3 above)
Comment (taylor-vann) [reply packages/labs/task/src/task.ts:200]: updated to not reset at all

File: packages/labs/task/src/task.ts (hunk 2 above)
Comment (taylor-vann) [reply packages/labs/task/src/task.ts:39]: updated

**Comment by taylor-vann** [comment]
@mohe2015 one potential use-case is lazy loading / lazy execution of a task, providing data for the initial render but before a remote request. In a compsci sense this would provide a kind of `Null Object` for tasks that correlate to an initial render.

As in, user loads page. Component needs a visual "stencil" up until the asynchronous task is requested / performed / completed.

**Review by sorvell — CHANGES_REQUESTED** [review]
Please add a test where `initialValue` is set.

**Comment by taylor-vann** [comment]
@sorvell added a test for initial values :) ready for another review

**Review by AndrewJakubowicz — CHANGES_REQUESTED** [review]
Please remove all the `.wireit` directory files.

They should have been ignored by `.gitignore`?

**Comment by taylor-vann** [comment]
thanks @AndrewJakubowicz, theyve been removed.

**Review by sorvell — CHANGES_REQUESTED** [review]
Almost there, just a couple of points to address, see comments. Let's also make sure to update the PR description to say that the value is only updated when the task completes (currently, it incorrectly says that it's reset to the initialValue).

File: .changeset/fifty-games-eat.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit-labs/task': patch
+---
+
+allow initial value
```
[the remark this answers is not in the document]
Comment (taylor-vann) [reply .changeset/fifty-games-eat.md:5]: yes! sorry, @AndrewJakubowicz reminded me and i forgot to change this PR as well, thank youu!

File: .changeset/fifty-games-eat.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit-labs/task': patch
```
[the remark this answers is not in the document]
Comment (taylor-vann) [reply .changeset/fifty-games-eat.md:2]: the change felt small but you're right, this is a feature addition. sounds good to me!

**Comment by taylor-vann** [comment]
Sweet! thanks yall
Andrew:
- got rid of the `.wireit` files
- left a more descriptive changelog
Steve:
- updated the description of the PR to correctly reflect behavior
- added a test to test the described behavior: only set `initialValue` on first `PENDING` state, not after.

**Comment by sorvell** [comment]
> `When a Task is "re-run", value is set as undefined.`

This is not true is it?

File: .changeset/lucky-apples-serve.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+'@lit-labs/task': minor
+---
+
+Added `initialValue` parameter to `TaskConfig`.
+Task can start with an initial value from TaskConfig.
```
Comment (sorvell) [inline .changeset/lucky-apples-serve.md:6]: We are changing the behavior of how value is treated when the task is re-run. This should be called out specifically here.

File: packages/labs/task/src/task.ts [hunk 7]
```diff
@@ -194,7 +196,6 @@ export class Task<T extends [...unknown[]] = any, R = any> {
     }
     this.status = TaskStatus.PENDING;
     this._error = undefined;
-    this._value = undefined;
```
Comment (sorvell) [inline packages/labs/task/src/task.ts:197]: This is the significant change that I want to make sure we're documenting and testing.

File: packages/labs/task/src/test/task_test.ts [hunk 8]
```diff
@@ -113,6 +113,17 @@ suite('Task', () => {
     assert.equal(el.taskValue, undefined);
   });
 
+  test('tasks start with initialValues in the first PENDING state', async () => {
```
[the remark this answers is not in the document]
Comment (sorvell) [reply packages/labs/task/src/test/task_test.ts:116]: We want specifically a test that validates that the value is *not* reset when the task is re-run (before it completes). This has nothing to do with initial value, but the behavior is being changed with this PR.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/labs/task/src/task.ts (hunk 7 above)
Comment (justinfagnani) [reply packages/labs/task/src/task.ts:197]: First I think we need to decide that this is the behavior we want. Is it required to add an initial value? Should it be controlled with an option? Let's chat.

**Review by e111077 — APPROVED** [review]
Technically this looks fine to me. Code LGTM

I'm also in favor of this behavior – Task has the tendency to flicker. Would be cool if `value` is the way to make it not

**Comment by justinfagnani** [comment]
Superseded by https://github.com/lit/lit/pull/4001

**Pull request closed without merging**